File: jar/src/main/kotlin/jline/lang/layered/CacheTask.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */

package jline.lang.layered;

import jline.lang.constant.ReplacementStrategy;
import jline.lang.constant.SchedStrategy;
import jline.lang.processes.Distribution;
import jline.lang.processes.Immediate;

/**
 * A task that offers caching services
 */
public class CacheTask extends Task {
    protected int items;
    protected int[] itemLevelCap;  // Changed to array to support multi-level caches
    protected ReplacementStrategy replacestrategy;

    public CacheTask(LayeredNetwork model, String name) {
        super(model, name);
        this.items = 1;
        this.itemLevelCap = new int[]{1};
        this.replacestrategy = ReplacementStrategy.FIFO; // Default replacement strategy
    }

    // Backward compatible constructor with single int capacity
    public CacheTask(LayeredNetwork model, String name, int items, int itemLevelCap, ReplacementStrategy replacestrategy) {
        super(model, name);
        this.items = items;
        this.itemLevelCap = new int[]{itemLevelCap};  // Convert single value to array
        this.replacestrategy = replacestrategy;
    }

    // New constructor supporting multi-level cache with int[] capacity
    public CacheTask(LayeredNetwork model, String name, int items, int[] itemLevelCap, ReplacementStrategy replacestrategy) {
        super(model, name);
        this.items = items;
        this.itemLevelCap = itemLevelCap;
        this.replacestrategy = replacestrategy;
    }

    // Backward compatible constructor with single int capacity
    public CacheTask(LayeredNetwork model, String name, int items, int itemLevelCap, ReplacementStrategy replacestrategy,
                     int multiplicity, SchedStrategy scheduling) {
        super(model, name, multiplicity, scheduling);
        this.items = items;
        this.itemLevelCap = new int[]{itemLevelCap};  // Convert single value to array
        this.replacestrategy = replacestrategy;
    }

    // New constructor supporting multi-level cache with int[] capacity
    public CacheTask(LayeredNetwork model, String name, int items, int[] itemLevelCap, ReplacementStrategy replacestrategy,
                     int multiplicity, SchedStrategy scheduling) {
        super(model, name, multiplicity, scheduling);
        this.items = items;
        this.itemLevelCap = itemLevelCap;
        this.replacestrategy = replacestrategy;
    }

    // Backward compatible constructor with single int capacity
    public CacheTask(LayeredNetwork model, String name, int items, int itemLevelCap, ReplacementStrategy replacestrategy,
                     int multiplicity, SchedStrategy scheduling, Distribution thinkTime) {
        super(model, name, multiplicity, scheduling, thinkTime);
        this.items = items;
        this.itemLevelCap = new int[]{itemLevelCap};  // Convert single value to array
        this.replacestrategy = replacestrategy;
    }

    // New constructor supporting multi-level cache with int[] capacity
    public CacheTask(LayeredNetwork model, String name, int items, int[] itemLevelCap, ReplacementStrategy replacestrategy,
                     int multiplicity, SchedStrategy scheduling, Distribution thinkTime) {
        super(model, name, multiplicity, scheduling, thinkTime);
        this.items = items;
        this.itemLevelCap = itemLevelCap;
        this.replacestrategy = replacestrategy;
    }

    // Backward compatible constructor for Python/user convenience
    public CacheTask(LayeredNetwork model, String name, int items, int itemLevelCap,
                     ReplacementStrategy replacementStrategy, int multiplicity) {
        super(model, name, multiplicity, SchedStrategy.FCFS);
        this.items = items;
        this.itemLevelCap = new int[]{itemLevelCap};  // Convert single value to array
        this.replacestrategy = replacementStrategy;
    }

    // New constructor for Python/user convenience with multi-level support
    public CacheTask(LayeredNetwork model, String name, int items, int[] itemLevelCap,
                     ReplacementStrategy replacementStrategy, int multiplicity) {
        super(model, name, multiplicity, SchedStrategy.FCFS);
        this.items = items;
        this.itemLevelCap = itemLevelCap;
        this.replacestrategy = replacementStrategy;
    }

    public int getItems() {
        return items;
    }

    public void setItems(int items) {
        this.items = items;
    }

    public int[] getItemLevelCap() {
        return itemLevelCap;
    }

    // Backward compatible getter returning total capacity (sum of all levels)
    public int getTotalCapacity() {
        int total = 0;
        for (int cap : itemLevelCap) {
            total += cap;
        }
        return total;
    }

    public void setItemLevelCap(int[] itemLevelCap) {
        this.itemLevelCap = itemLevelCap;
    }

    // Backward compatible setter with single int
    public void setItemLevelCap(int itemLevelCap) {
        this.itemLevelCap = new int[]{itemLevelCap};
    }

    public ReplacementStrategy getReplacestrategy() {
        return replacestrategy;
    }

    public void setReplacestrategy(ReplacementStrategy replacestrategy) {
        this.replacestrategy = replacestrategy;
    }
}

File: matlab/src/lang/layered/CacheTask.m
classdef CacheTask < Task
    % A software server in a LayeredNetwork.
    %
    % Supports multi-level cache hierarchies with configurable capacities
    % per level (e.g., L1, L2, L3 caches).
    %
    % Copyright (c) 2012-2026, Imperial College London
    % All rights reserved.

    properties

        items;
        itemLevelCap;  % Scalar or array for multi-level cache capacities
        replacestrategy;
        retrieval = false;  % delayed-hit retrieval on the miss path

    end

    methods
        %public methods, including constructor

        %constructor
        function self = CacheTask(model, name, nitems, itemLevelCap, replStrat, multiplicity, scheduling)
            %self = CacheTask(model, name, nitems, itemLevelCap, replStrat, multiplicity, scheduling)
            %
            % itemLevelCap can be:
            %   - Scalar: Single-level cache with given capacity
            %   - Vector: Multi-level cache with capacity per level

            if ~exist('name','var')
                line_error(mfilename,'Constructor requires to specify at least a name.');
            end

            if nargin < 6
                multiplicity = 1;
            end

            if nargin < 7
                scheduling = SchedStrategy.FCFS;
            end
            self@Task(model, name, multiplicity, scheduling);

            self.items = nitems;

            % Validate and store itemLevelCap (scalar or array)
            if isscalar(itemLevelCap)
                self.itemLevelCap = itemLevelCap;
            elseif isvector(itemLevelCap)
                self.itemLevelCap = itemLevelCap(:)';  % Ensure row vector
            else
                line_error(mfilename, 'itemLevelCap must be scalar or vector');
            end

            self.replacestrategy = replStrat;
        end

        % Get item level capacity
        function cap = getItemLevelCap(self, level)
            % CAP = GETITEMLEVELCAP(SELF, LEVEL)
            %
            % Get capacity for a specific cache level or all levels.
            %
            % Args:
            %   level: (optional) Cache level index (1-based)
            %
            % Returns:
            %   cap: Capacity value(s) - scalar if level specified, array otherwise

            if nargin == 1
                % Return all levels
                cap = self.itemLevelCap;
            else
                % Return specific level
                if isscalar(self.itemLevelCap)
                    % Single-level cache
                    if level ~= 1
                        line_error(mfilename, 'Single-level cache only has level 1');
                    end
                    cap = self.itemLevelCap;
                else
                    % Multi-level cache
                    if level < 1 || level > length(self.itemLevelCap)
                        line_error(mfilename, sprintf('Level %d out of range [1, %d]', level, length(self.itemLevelCap)));
                    end
                    cap = self.itemLevelCap(level);
                end
            end
        end

        % Get total capacity across all levels
        function total = getTotalCapacity(self)
            % TOTAL = GETTOTALCAPACITY(SELF)
            %
            % Get the sum of capacities across all cache levels.
            %
            % Returns:
            %   total: Sum of all level capacities

            total = sum(self.itemLevelCap);
        end

        % Enable/disable delayed-hit retrieval on the miss path
        function self = setRetrieval(self, retrieval)
            % SETRETRIEVAL(SELF, RETRIEVAL)
            %
            % Enable a retrieval system with delayed-hit coalescing on the cache
            % miss path. When set, concurrent misses for the same item arriving
            % while a fetch (the miss-branch activity and its backend calls) is in
            % flight are parked and released together as delayed hits when the
            % fetch completes, instead of each triggering an independent fetch.
            if nargin < 2
                retrieval = true;
            end
            self.retrieval = logical(retrieval);
        end

        function tf = hasRetrieval(self)
            tf = self.retrieval;
        end

        % Set item level capacity
        function setItemLevelCap(self, caps)
            % SETITEMLEVELCAP(SELF, CAPS)
            %
            % Set cache level capacities.
            %
            % Args:
            %   caps: Scalar for single-level or vector for multi-level

            if isscalar(caps)
                self.itemLevelCap = caps;
            elseif isvector(caps)
                self.itemLevelCap = caps(:)';  % Ensure row vector
            else
                line_error(mfilename, 'caps must be scalar or vector');
            end
        end

        % Get number of cache levels
        function n = getNumberOfLevels(self)
            % N = GETNUMBEROFLEVELS(SELF)
            %
            % Get the number of cache levels.
            %
            % Returns:
            %   n: Number of levels (1 for single-level, >1 for multi-level)

            if isscalar(self.itemLevelCap)
                n = 1;
            else
                n = length(self.itemLevelCap);
            end
        end
    end
end